[flang][OpenMP] Support custom mappers in target update to/from clauses

KrxGu opened the pull request:
Fixes #168701

This patch adds support for custom mappers in OpenMP `target update` `to()` and `from()` clauses, implementing the OpenMP 5.2 specification.

Previously, custom mappers were only supported in `map` clauses. This change extends mapper resolution to `target update` constructs by:

- Adding `Pre(OmpClause::To)` and `Pre(OmpClause::From)` handlers in `resolve-names.cpp`
- Following the same pattern as the existing `OmpMapClause` handler
- Supporting both custom mappers and the special `default` mapper
- Adding proper diagnostics for undefined mappers and non-mapper symbols

**Testing:**
- Semantic tests verify custom/default mapper usage and error detection
- Lowering tests verify MLIR generation with mapper attributes
<img width="697" height="202" alt="image" src="https://github.com/user-attachments/assets/b081e118-08ba-4b74-8d61-c313aec4811f" />

Commit:
[flang][OpenMP] Support mappers in target update to/from

Fixes #168701

Commit:
[flang][OpenMP] Implement mapper lowering for target update

Support mapper modifier in processMotionClauses for to/from clauses

TIFitis commented on flang/lib/Semantics/resolve-names.cpp:
```diff
@@ -1876,6 +1878,64 @@ bool OmpVisitor::Pre(const parser::OmpMapClause &x) {
+      }
+    }
+  }
+  return true;
```
Have you considered this suggestion?

TIFitis reviewed (commented):
Thanks for the PR. I've added some review comments below...

Commit:
[flang][OpenMP] Refactor mapper resolution per review

Extract ResolveMapperModifier helper and simplify tests

Commit:
[flang] Fix clang-format issues

Commit:
[flang] Fix test expectations for program-scoped mapper names

Commit:
[flang] Extract mapper resolution into helper function

KrxGu commented:
@kparzysz , @TIFitis would love to get this merged!!

skatrak reviewed (commented):
Thank you for this work. I have some minor comments, but I'll leave approval to @TIFitis and @kparzysz.

KrxGu committed:
Refactor mapper code based on review feedback

- Use early return in getMapperIdentifier to reduce nesting
- Remove redundant if constexpr with identical branches
- Make ResolveMapperModifier parameter const
- Remove unnecessary const_cast at call sites

KrxGu committed:
Remove unnecessary const_cast for mutable symbol member

The symbol field is already marked mutable in the Name struct,
so it can be modified directly through a const reference.

KrxGu committed:
Add offloading test for target update with custom mappers

Use separate objects (obj1, obj2) for custom and default mapper
tests to avoid device memory mapping conflicts.

KrxGu committed:
Fix parser API usage: type.derived.t -> type.v.t

DeclarationTypeSpec::Type is a WRAPPER_CLASS with member 'v',
not 'derived'. Corrects build failure on CI.

TIFitis reviewed (approved):
LGTM overall minus the nit. Please allow another +1.

The offload test is passing on my end, thanks for the PR.

KrxGu committed:
Remove unrelated formatting changes from rebase

Per review feedback, removed all pointer-style and indentation
changes that were picked up during rebase. Kept only functional
changes for mapper support in target update.
